### DISCUSSION TITLE
ULCL of free5GC SMF changed

### DISCUSSION THREAD

**Discussion by s5uishida** [nextmn/.github#1] [General]
Hi @louisroyer 

According to [this](https://github.com/free5gc/free5gc/issues/599#issuecomment-2357448799), the ULCL code of free5GC SMF has been changed.

As a result, NextMN-UPF's ULCL have not worked.

FYI.

**Comment by louisroyer** [comment]
Thank you very much for sharing this information with me, and for the issue reporting on Free5GC :)

If I understand correctly, this issue is now fixed in the SMF?

Have you tested the new version with NextMN-UPF yet? If it doesn't work, please let me know.

**Reply by s5uishida** [reply]
Hi @louisroyer 

ULCL did not work with this modified free5GC SMF and the latest NextMN-UPF.

With the SMF prior to this fix, ULCL worked in combination with the latest NextMF-UPF.

Best regards,

--Shigeru

**Reply by louisroyer** [reply]
Sorry for misunderstanding your first message. I thought it didn't work before the fix_ instead of _doesn't work since the fix_.

This should now work (with v0.0.10).

Thank you again.

**Reply by s5uishida** [reply]
Hi @louisroyer 

Thank you very much for your efforts.

I tried the following scenarios using the latest versions of free5GC and NextMN-UPF, but ULCL did not work.

https://github.com/s5uishida/free5gc_ueransim_ulcl_sample_config

For pings to the following destinations, all packets were passed through PSA-UPF.
This is not the expected result.

- google.com
- 8.8.8.8
- 172.17.0.1

Also I have confirmed that ULCL correctly using the latest free5GC (C-Plane) and the latest free5GC's go-upf (UPF).

FYI.

**Reply by s5uishida** [reply]
I updated free5GC as follows and built it.
```
git clone --recursive -j `nproc` https://github.com/free5gc/free5gc.git
cd ~/free5gc
git checkout main
git submodule sync
git submodule update --init --jobs `nproc`
git submodule foreach git checkout main
git submodule foreach git pull --jobs `nproc`
```

**Reply by louisroyer** [reply]
Hi @s5uishida,
I can reproduce on my side with the following setup:
```text
  specificPath:
    - dest: 10.4.0.1/32
      path: [UPFI, UPFA1]
    - dest: 10.4.0.2/32
      path: [UPFI, UPFA2]
```
(my UE is 10.2.1.129)

PDR created by the latest version of the SMF are strange (see [PCAP attached](https://github.com/user-attachments/files/17056386/pfcp-f5gc.pcapng.tar.gz)):
- In packet number 41, the source interface is Access, and the flow description is `permit out ip from 10.4.0.1/32 to 10.2.1.129` : this is correct
- In packet number 53, the source interface is Access, and the flow description is `permit out ip from 10.2.1.129 to 10.4.0.2/32` : this is not correct (source and destination in wrong order).

[`UpdateRANAndIUPFUpLink`](https://github.com/lyz508/smf/blob/5c2a162c3ecfcec0bcd23dfceafe58005df93237/internal/sbi/processor/ulcl_procedure.go#L288) probably has the same bug that was fixed in `EstablishULCL`.

ULCL happens to work correctly with free5gc/go-upf, and I cant figure out by what miracle : with the current PFCP rules this PDR should never be matched ; there is probably another bug in free5gc/go-upf.

Can you attach me a PCAP on N4 interface to confirm this is the exact same issue? Or logs from NextMN-UPF (it displays current rules periodically).

**Reply by louisroyer** [reply]
<s>I might have found why free5gc/go-upf still works with wrong rules:
When the source interface is Access, it swap source and destination to convert filter description in this way:
https://github.com/free5gc/go-upf/blob/922281ba58b6ca42d73be218671c8feae3658420/internal/forwarder/gtp5g.go#L182-L185
but here, the `fd.Dir` must also be reversed (`out`/`in`) to work correctly.

This means the wrongly created `permit out ip from 10.2.1.129 to 10.4.0.2/32` matches, but `permit out ip from 10.4.0.1/32 to 10.2.1.129` does not. But there is a third "default" rule without SDF Filter that is matched instead, that have an identical FAR.

I will test this and, if it works, send pull requests to Free5GC.</s>

EDIT: there is no bug in free5gc/go-upf, and the inversion in modified pdr is not related (but still a bug). The direct cause is a mistake I made while patching.

Version 0.0.11 should work.

**Reply by s5uishida** [reply]
Hi @louisroyer 

It worked fine with v0.0.11.

Thank you very much!

**Reply by s5uishida** [reply]
Hi @louisroyer 

I found your [PR](https://github.com/free5gc/smf/pull/126) for free5GC SMF, so I tried it.

As a result, ULCL worked fine.

Thank you very much for your efforts!
